### PR TITLE
Bag of stuff.

### PR DESCRIPTION
**Pull request by markhibberd**
@thumphries I am going to put up some PRs to remove the submodule dependencies so it gets to a publishable to hackage state. But there are a bunch of commits I am working on top of that make it pretty hard to make changes if they are not in your repository. They are things I have done to just get through, so may not be the neatest - but are all notional improvements/fixes - can change them as required to get them in. Sorry it is a bit of a mess - hold onto a fork for a year and a few cobwebs build up.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -1,18 +1,18 @@
 [submodule "lib/p"]
 	path = lib/p
-	url = git@github.com:ambiata/p
+	url = git@github.com:markhibberd/p
```
Comment (markhibberd) [inline .gitmodules]: Submodule changes are probably less than ideal, but depend on 8.2/8.4 fixes from my forks. Going to try and snip these out on subsequent PRs anyway so may not matter, but if you want to pull in the master branches of those forks in some way, I could re-point to your forks if needed.

**Comment by markhibberd** [comment]
Running some builds - https://circleci.com/gh/markhibberd/projector/tree/master - a lot of this stuff hasn't left my machine - so this might be interesting.

File: .gitmodules (hunk 1 above)
Comment (thumphries) [reply .gitmodules]: This is fine by me - anything that keeps this stuff building is incredibly valuable. I'll give you push so we can keep it ticking over.

My eventual preference is to inline some of these ambiata libraries where necessary so nobody has to publish them. Started on this last year in https://github.com/thumphries/projector/compare/topic/deambiata but got burned out and stopped.

**Comment by thumphries** [comment]
All good changes - have looked these commits over already, so if you manage to get it to pass CI I'm happy for you to merge.

**Comment by markhibberd** [comment]
https://circleci.com/workflow-run/95873d91-2529-4bcf-9ad9-39c853012cd8

**Comment by markhibberd** [comment]
@thumphries This is ready to go, but I don't think you gave me push access so can't merge.

**Pull request merged**